src/Symfony/Component/Security/Http/Firewall/ContextListener.php

```diff
@@ -219,12 +220,19 @@ protected function refreshUser(TokenInterface $token)
                 $userNotFoundByProvider = true;
             }
         }
-
         if ($userDeauthenticated) {
             if (null !== $this->logger) {
                 $this->logger->debug('Token was deauthenticated after trying to refresh it.');
             }
 
+            if (null !== $this->dispatcher) {
+                $this->dispatcher->dispatch(new UserHasChangedEvent($token, $user), UserHasChangedEvent::class);
+
+                if ($token->isAuthenticated()) {
```

Comment:
it must call `$token->setUser($refreshedUser);`, which will makes your test fail below (and then you can fix it by reversing the assertion)